### PR TITLE
IGNITE-19362 Fix deadlock while restarting with existing raft snapshot and scheduled index rebuild

### PR DESCRIPTION
**Pull request by SammyVimes**
https://issues.apache.org/jira/browse/IGNITE-19362

### CODE REVIEW

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java [hunk 1]
```diff
@@ -878,6 +879,48 @@ public void nodeWithDataTest() throws InterruptedException {
...
+
+        createTableWithData(List.of(ignite), TABLE_NAME, 1, 20);
+
+        TableImpl table = (TableImpl) ignite.tables().table(TABLE_NAME);
+
+        InternalTableImpl internalTable = (InternalTableImpl) table.internalTable();
+
+        for (int i = 0; i < 20; i++) {
+            // Flush data on disk, so that we will have a snapshot to read on restart.
+            internalTable.storage().getMvPartition(i).flush().join();
```
Comment (sashapolo) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java:897]: Let's not use `join` here and use an assertion instead

Comment (sashapolo) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java:897]: Also, these flushes can be done in parallel, no need to wait for them sequentially

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java [hunk 2]
```diff
@@ -878,6 +879,48 @@ public void nodeWithDataTest() throws InterruptedException {
...
+        }
+
+        // Add more data, so that on restart there will be a index rebuilding operation.
+        try (Session session = ignite.sql().createSession()) {
+            for (int i = 0; i < 100; i++) {
+                session.execute(null, "INSERT INTO " + TABLE_NAME + "(id, name) VALUES (?, ?)",
+                        i + 500, VALUE_PRODUCER.apply(i + 500));
```
Comment (sashapolo) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java:904]: What's `+ 500` for?

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/snapshot/SnapshotExecutorImpl.java [hunk 3]
```diff
@@ -250,7 +250,13 @@ public boolean init(final SnapshotExecutorOptions opts) {
         final FirstSnapshotLoadDone done = new FirstSnapshotLoadDone(reader);
         Requires.requireTrue(this.fsmCaller.onSnapshotLoad(done));
         try {
-            done.waitForRun();
+            // TODO: IGNITE-19363 We want to avoid deadlock for now, but this is an ad-hoc decision.
+            // We don't wait for the partition's snapshot load closure to finish here.
+            if (!node.getNodeId().getGroupId().contains("part")) {
+                done.waitForRun();
+            } else {
+                done.status = new Status();
```
Comment (sashapolo) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/snapshot/SnapshotExecutorImpl.java:258]: How does this work? Don't we have a race here with setting the task status? What will happen if snapshot loading fails?

Comment (sashapolo) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/snapshot/SnapshotExecutorImpl.java:258]: Also, there's a `Status.OK()` method, which looks a little bit more descriptive

Comment (SammyVimes) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/snapshot/SnapshotExecutorImpl.java:258]: This is a temporal solution, so we don't really care about failing loading the snapshot. The race doesn't matter, we won't be checking the result

Comment (sashapolo) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/snapshot/SnapshotExecutorImpl.java]: I really hope that it is temporary, not checking the status seems dangerous

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java [hunk 4]
```diff
@@ -878,6 +879,54 @@ public void nodeWithDataTest() throws InterruptedException {
...
+
+        ignite = startNode(0);
+
+        checkTableWithData(ignite, TABLE_NAME);
+
+        table = (TableImpl) ignite.tables().table(TABLE_NAME);
+
+        // Check data that was added after flush.
+        for (int i = 0; i < 100; i++) {
+            Tuple row = table.keyValueView().get(null, Tuple.create().set("id", i));
```
Comment (sashapolo) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java:924]: How does this work now, if we still insert data with id equal to `i + 500`?

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/snapshot/SnapshotExecutorImpl.java (hunk 3 above)
Comment (SammyVimes) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/snapshot/SnapshotExecutorImpl.java]: It is, I am going to start on fixing it right away

**Review by sashapolo — APPROVED** [review]

**Pull request merged**